Step2: 계산기 리뷰 요청 드립니다 :)

kwanguuuu opened the pull request:
안녕하세요. 리뷰어님. 
step 2 리뷰 요청 드립니다 :)

kwanguuuu committed:
[step2] feat: 문자열 parse 테스트 작성

kwanguuuu committed:
[step2] fix: 패키지 변경 ( step 별로 구분 )

kwanguuuu committed:
[step2] feat: 공백 문자열 제거 테스트 추가

kwanguuuu committed:
[step2] feat: 연산자 enum class 테스트, 클래스 작성

kwanguuuu committed:
[step2] feat: 계산기 1차 작성, 테스트 작성

kwanguuuu committed:
[step2] feat: 계산기 작성

kwanguuuu committed:
[step2] refactor: parser refactoring

kwanguuuu committed:
[step2] add main fun

kwanguuuu committed:
[step2] fix: gradlew ktlintFormat 설정

wisemuji commented on src/main/kotlin/step2/calculator/Operator.kt:
```diff
@@ -0,0 +1,36 @@
+    PLUS("+", { a, b -> a + b }),
+    MINUS("-", { a, b -> a - b }),
+    MULTIPLY("*", { a, b -> a * b }),
+    DIVIDED("/", { a, b -> a / b })
```
a, b 나누기 결과의 소수점 아래 값이 유실되고 있어요. 유실되지 않도록 저장해보면 어떨까요?
<img width="313" alt="image" src="https://user-images.githubusercontent.com/32327475/166856355-ca884646-bf14-4d7b-b167-4b97c8898656.png">

kwanguuuu replied:
말씀하신것 처럼  소숫점 처리를 제대로 못했군요 ...😇
Long type으로 처리하던걸 double로 변경했습니다. 관련된 타입과 메서드명 리팩토링 적용반영했습니다.

https://github.com/next-step/kotlin-racingcar/pull/431/commits/89e8b9c46f5c9227435c0125a560347760c9628e

wisemuji commented on src/main/kotlin/step2/calculator/Calculator.kt:
```diff
@@ -0,0 +1,43 @@
+        return Operator.getOperator(token[OPERATOR_POSITION_INDEX])
+    }
+
+    companion object {
+        private const val OPERATOR_POSITION_INDEX = 1
+
+        private const val FIRST_NUMBER_POSITION_INDEX = 0
+        private const val SECOND_NUMBER_POSITION_INDEX = 2
+
+        private const val NEXT_OPERATOR_START_INDEX = 3
+    }
```
매직 넘버를 상수로 잘 분리해주셨네요 👍

wisemuji commented on src/main/kotlin/step2/calculator/Operator.kt:
```diff
@@ -0,0 +1,36 @@
+
+enum class Operator(
+    private val symbol: String,
+    private val function: BiFunction<Long, Long, Long>
```
BiFunction도 좋지만 코틀린의 람다 표현식을 활용해보면 어떨까요? 
```kotlin
private val function: (Long, Long) -> Long
```

이렇게 선언하면 호출부에서 다음과 같이 간단하게 작성할 수 있어요.
```kotlin
return function(first.toLong(), second.toLong())
```

kwanguuuu replied:
감사합니다. 람다식을 직접 사용하는게 어색하니. 동작이 가능하게 일단 작성했던 것 같아요 ㅎ
pr 반영했습니다.
https://github.com/next-step/kotlin-racingcar/pull/431/commits/8158b35eaf108c1c2a6a9843d37e2ace932658cd

wisemuji commented on src/main/kotlin/step2/calculator/Operator.kt:
```diff
@@ -0,0 +1,36 @@
+        } catch (cause: NumberFormatException) {
+            throw IllegalArgumentException("숫자가 아닙니다. cause = `$value`")
+        }
+    }
```
이 역할/책임을 Operator가 들고 있는게 맞을까요?

kwanguuuu replied:
operator의 역할은 아닌것 같구요, 따로 검증 클래스가 있어야 하는지 고민을 하다가 Operator에 있어도 될것 같단 생각에 일단 작성했는데요,
해당 리뷰 + 다른 리뷰를 통해 코드 작성하다보니
NumberPair라는 객체를 따로 도출했는데요, 역할이 거기로 넘어갔습니다.! 제 생각엔 괜찮은것 같은데 리뷰어님이 한번 봐주시면 좋겠어요!
https://github.com/next-step/kotlin-racingcar/pull/431/commits/8e42aefb3e30b5f453ba98dcee37bff72b8becaa

wisemuji replied:
NumberPair 일급 객체로 래핑하신 방향 좋습니다 👍 일급 객체는 앞으로 과정을 계속 수강하시다 보면 더 자세히 다루게 될거예요. 지금 분리하신 것처럼 객체가 적절하지 못한, 혹은 너무 많은 역할/책임을 들고 있는 것으로 보이면 새로운 객체를 만들거나 더 적절해보이는 객체로 위임하는걸 고려해보세요!

wisemuji replied:
[일급 컬렉션을 사용하는 이유](https://tecoble.techcourse.co.kr/post/2020-05-08-First-Class-Collection/)
이 글도 참고하시면 좋습니다 👍

wisemuji commented on src/test/kotlin/step2/calculator/OperatorTest.kt:
```diff
@@ -0,0 +1,23 @@
+        val test2 = "-"
+        val test3 = "*"
+        val test4 = "/"
+        val test5 = "#"
```
테스트 코드를 작성할 때에도 유지보수가 쉽도록 코드 퀄리티를 고려해야 합니다. 간단한 테스트용 변수명을 지을 때에도 다른 개발자들이 알아볼 수 있도록 의미를 담아보면 어떨까요?

kwanguuuu replied:
ㅎㅎ 감사합니다! 수정했습니다. 다른 테스트 메서드의 변수명들도 봤는데.. 뭔가 의미를 도출하기엔 애매한것 같아서,
리뷰주신 코드만 수정했습니다.

wisemuji commented on src/main/kotlin/step2/calculator/Calculator.kt:
```diff
@@ -0,0 +1,43 @@
+
+    private fun _calculate(token: List<String>): Long {
+        val operator = getOperator(token)
+        val (first, second) = Pair(token[FIRST_NUMBER_POSITION_INDEX], token[SECOND_NUMBER_POSITION_INDEX])
```
Kotlin `to` 키워드를 사용하시면 Pair를 생략하실 수 있습니다.
참고: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/to.html

kwanguuuu replied:
음 이부분도 리뷰 제안 주셨던 것 처럼
Long to Long 으로 pair 객체를 생성했는데요,
NumberPair일급 객체로 역할을 변경하다보니 outdate됐습니다!
-  https://github.com/next-step/kotlin-racingcar/pull/431#discussion_r867483758 
이 리뷰랑 함꼐 봐주시면 좋을것 같아요!

kwanguuuu replied:
+추가적으로 어색한 코드작성이 있는데요,
private 함수 작성시 컨벤션으로 언더바`_calculate` 를 가져갔는데,
이유는 calculate.. 이상의 적절한 함수명을 찾지 못했고, public calculate가 이미 존재하고 있어서입니닷.

물론 함수 파라미터가 달라서 오버라이딩 할 수도 있었는데, 구분이 있으면 더 좋지 않을까? 란 생각이었어요.

리뷰어님은 어떻게 생각하시나요??

wisemuji replied:
Kotlin coding convention 가이드에서는 [backing property 변수에 언더스코어가 들어가는 경우](https://kotlinlang.org/docs/coding-conventions.html#names-for-backing-properties)가 있긴 하지만 [함수 이름에 언더스코어를 넣는 건 지양하라고](https://kotlinlang.org/docs/coding-conventions.html#function-names) 합니다. 

이 경우에는 말씀해주신것처럼 오버로딩 하는 것도 좋은 방법이라고 생각합니다. 조오금만 더 나아가보자면 현재 Calculator 클래스는 raw한 string을 변환하는 역할, 계산하는 로직을 모두 들고 있는데요, Calculator 클래스에 raw한 string을 바로 넘기기보다 Operator, Number 객체로 1차 변환한 후, 그 객체로 연산하는 역할만 Calculator에서 들고 있게 하는 방법도 있을 것 같아요.

관련해서는 제가 위에 남긴 객체의 역할/책임과 관련된 코멘트도 참고해보시면 좋을 것 같습니다.

> 객체가 적절하지 못한, 혹은 너무 많은 역할/책임을 들고 있는 것으로 보이면 새로운 객체를 만들거나 더 적절해보이는 객체로 위임하는걸 고려해보세요!

wisemuji commented on src/main/kotlin/step2/calculator/Calculator.kt:
```diff
@@ -0,0 +1,43 @@
+
+    private fun getOperator(token: List<String>): Operator {
+        return Operator.getOperator(token[OPERATOR_POSITION_INDEX])
+    }
```
`getOperator`의 함수명만 보고 내부에서 `Operator.getOperator(token[OPERATOR_POSITION_INDEX])` 의 동작을 할 것이라고 예상할 수 있을까요? 어떤 함수명이 다른 개발자도 쉽게 이해할 수 있는 함수명일까요?

kwanguuuu replied:
Calculator안에 getOperator란 프라이빗 메서드가 Operator를 호출하는 모양인데요,
리뷰어님 의견중에 궁금한건..
- getOperator의 메서드명이 혼란을 주는지
- getOperator안에 Operator를 호출하는것이 혼란이 주는지

가 궁금합니닷!
메서드명은 사실 getOperator보단 findOperator가 올바른 표현인것 같아서, 변경했구요
굳이 getOperator란 프라이빗 함수를 호출할 필요 없이.
Operator.getOperator 메서드를 calculate() 함수에서 직접 호출해도 될것 같단 생각이 들었습니다.!

이부분 의견 궁금합니당..

wisemuji replied:
getOperator의 메서드명이 getOperator가 하는 일을 전부 드러내고 있지 않아 드린 코멘트였습니다!

제가 이해한 것이 맞다면 getOperator은 parameter인 token의 OPERATOR_POSITION_INDEX에 위치한 element에 해당하는 Operater를 가져오는 역할을 하는 것 같은데요, 

```kotlin
private fun getOperator(token: List<String>): Operator
```

호출부에서만 보면 내부에서 어떤 로직을 통해 token 리스트 중에 하나의 Operater만 가져오고 있는지 이해하기 어렵습니다. 

말씀해주신 방향처럼 calculate() 함수에서 직접 호출하는 것도 좋은 구현 방향이라고 생각하지만 전체 token 리스트 중에서 현재 계산에서만 사용할 numberPair와 operator를 가져온다는걸 더 명확하게 하려면 함수를 활용하는게 좋지 않을까 하는 의견입니다!

wisemuji commented on src/main/kotlin/step2/calculator/utils/CustomStringParser.kt:
```diff
@@ -0,0 +1,20 @@
+
+        private const val DELIMITER = " "
+    }
+}
```
companion object만 존재하는 클래스의 경우 object 타입으로 만들어보면 어떨까요?
```
object CustomStringParser
```

kwanguuuu replied:
수정했습니다 :). 
kotlin 컨벤션의 순서가 같은 메모리 영역이면 
- 변수
- 함수
순서 인것 같아 DILIMITER 상수 위치도 변경했는데 확인 부탁드릴게요!

wisemuji replied:
잘 변경해주셨습니다 👍 추가로 kotlin class layout convention도 참고하시면 더욱 도움이 되실 것 같아요.

https://kotlinlang.org/docs/coding-conventions.html#class-layout

wisemuji commented on src/test/kotlin/step2/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,24 @@
+class CalculatorTest {
+
+    @ParameterizedTest
+    @CsvSource(
+        value = [
+            "2 + 5 = 7",
+            "2 - 5 = -3",
+            "5 / 1 = 5",
+            "2 + 3 * 4 / 2 = 10"
+        ],
+        delimiter = '='
+    )
```
CsvSource를 이용해서 ParameterizedTest를 잘 구현해주셨군요 👍

wisemuji reviewed (changes requested):
Step2 미션을 잘 구현해주셨습니다! 🎉 

메서드가 너무 많은 일을 하지 않도록 분리하기 위해 노력하신 모습이 보입니다 👍
고민해보시면 좋을 코멘트를 달았습니다. (이해되지 않는게 있으시다면 DM으로 물어보세요!) 

또 기능 단위로 커밋을 하시면 더욱 좋습니다. 기능을 구현하기 전에 README.md 파일에 구현할 기능 목록을 정리해 추가한 후 이 기능 목록을 각각의 커밋으로 만들어보면 어떨까요? 현 시점에서 적용이 어렵다면 Step3부터 적용해주셔도 좋습니다. 🙂

kwanguuuu committed:
[step2] docs: readme.md 수정, 진행된 checkbox 확인

- pr 리뷰 반영 : https://github.com/next-step/kotlin-racingcar/pull/431#pullrequestreview-961481591

kwanguuuu committed:
[step2] bug fix: 나눗셈 소수점 누락되는 버그 수정, pr 반영

- pr 리뷰 반영 : https://github.com/next-step/kotlin-racingcar/pull/431#discussion_r864564175

kwanguuuu committed:
[step2] refactor: CustomStringParser.kt object 클래스로 변경

- pr 리뷰 반영 : https://github.com/next-step/kotlin-racingcar/pull/431#discussion_r864586106

kwanguuuu committed:
[step2] refactor: 메서드명 변경

- CustomStringParser.parse() -> CustomStringParser.splitAswords()
- 관련 로컬변수명 변경

kwanguuuu committed:
[step2] refactor: Pair 사용 방볍 변경

kwanguuuu committed:
[step2] test, refactor: Pair 대신 NumberPair 객체 생성, 생성 테스트 작성

kwanguuuu committed:
[step2] refactor: Operator, Calculator이 NumberPair 사용하도록 수정

kwanguuuu committed:
[step2] fix: Operator 메서드명 수정

 - getOperator -> findOperator
 - pr반영..? https://github.com/next-step/kotlin-racingcar/pull/431#discussion_r864584972

kwanguuuu committed:
[step2] chore: Operator test 변수명 수정

kwanguuuu committed:
[step2] test, fix: 코틀린 람다식 테스트, 자바 function 제거

- pr 반영: https://github.com/next-step/kotlin-racingcar/pull/431#discussion_r864571636

kwanguuuu commented:
@wisemuji 
리뷰어님 안녕하세요:) 연휴를 보내다 보니.. 주신 pr피드백이 늦었네요!
pr 주신 내용 모두 수정했는데 확인 부탁드리겠습니다 

+제가 추가로 궁금한걸 남겼는데.. 답변 주시면 감사하겠습니다 :) ( 미션 진행사항일정에 차질이 있다면..디엠으로 토론......해도 좋을것같아요..)

wisemuji commented on src/test/kotlin/step2/calculator/study/FunctionTest.kt:
```diff
@@ -0,0 +1,30 @@
+import org.junit.jupiter.api.Test
+import java.util.function.BiFunction
+
+class FunctionTest {
```
강의에서 배운 학습 테스트 작성을 바로 적용하시는 모습 👍 넘 좋습니다!

wisemuji commented on src/main/kotlin/step2/calculator/Calculator.kt:
```diff
@@ -0,0 +1,47 @@
+        return _calculate(token)
+    }
+
+    private fun _calculate(token: List<String>): Double {
+        val numberPair = getNumberPair(token)
+        val operator = getOperator(token)
+
+        val result = operator.operate(numberPair)
+
+        if (hasNextOperation(token)) {
+            val leftToken: List<String> = token.subList(NEXT_OPERATOR_START_INDEX, token.size)
+            return _calculate(listOf(result.toString()) + leftToken)
+        }
+        return result
+    }
```
나중에 재귀 함수를 사용하시게 되면 `tailrec` 키워드를 이용해 재귀를 최적화하는 방법을 고려해보세요!

https://kotlinlang.org/docs/functions.html#tail-recursive-functions

wisemuji commented on README.md:
```diff
@@ -1 +1,12 @@
-# kotlin-racingcar
\ No newline at end of file
+# kotlin-racingcar
+
+## step1 기능 요구사항 정리
+- [x] 입력한 문자열로 사칙연산을 수행할 수 있는 기능 구현하기
+  - [x] 입력값이 null 이거나 빈 공백 문자 인 경우 IllegalArgumentException
+  - [x] 사칙연산 기호가 아닌 경우 IllegalArgumentException
+    - 연산자 확인 기능 추가
+  - [x] 공백 문자열을 빈 공백 문자로 처리
+    - 빈 공백 문자 삭제 기능 추가
+    - space, tab, 줄 나눔 공백 삭제 기능 추가 
+    
+- [x] 문자열 계산은 사칙 연산 순서가 아닌, 입력 순서대로 작동하도록 하기
```
요구사항 정리 좋습니다! 💯 

미리 정리해두신 상태로 단위 테스트할때 활용하시면 테스트 범위를 정의하는데에 도움이 되실거예요.

wisemuji reviewed (approved):
리뷰 반영을 너무 잘 해주셨습니다!! 👏

어떻게 더 좋은 코드를 작성할 수 있을지 고민하신 흔적이 보입니다 💯  
남겨주신 질문에 대한 답변을 남겨두었으니 고민해보시고 다음 미션에 참고해주시면 좋을 것 같습니다.
남은 미션도 이대로 계속 화이팅입니닷!!!!

다음 자동차 경주 미션을 진행하기 전에 다음 글이 도움이 되면 좋겠습니다 🙏 
* [좋은 객체의 7가지 덕목](https://codingnuri.com/seven-virtues-of-good-object/)
* [객체지향의 사실과 오해 [5장. 책임과 메시지]](https://github.com/Coveong/reading-books-for-programmers/blob/main/%EA%B0%9D%EC%B2%B4%EC%A7%80%ED%96%A5%EC%9D%98%20%EC%82%AC%EC%8B%A4%EA%B3%BC%20%EC%98%A4%ED%95%B4/05%EC%9E%A5.%20%EC%B1%85%EC%9E%84%EA%B3%BC%20%EB%A9%94%EC%8B%9C%EC%A7%80/wisemuji.md)

궁금한 점이나 더 논의하고 싶은 내용이 있다면 언제든 편하게 DM 주세요!
